### PR TITLE
[Feature-WIP](inverted index) support use inverted index searcher cache

### PR DESCRIPTION
**Pull request by Tanya-W**
# Proposed changes
use inverted index searcher cache to improve query performance

dependency pr: https://github.com/apache/doris/pull/14211 https://github.com/apache/doris/pull/15807 https://github.com/apache/doris/pull/15823

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/common/config.h [hunk 1]
```diff
@@ -875,6 +875,16 @@ CONF_String(disposable_file_cache_path, "");
...
+
+// inverted index searcher cache
+// cache entry stay time after lookup, default 1h
+CONF_mInt32(index_cache_entry_stay_time_after_lookup_s, "3600");
+// inverted index searcher cache size
+CONF_String(inverted_index_searcher_cache_limit, "10%");
+// set `true` to enable insert searcher into cache when write inverted index data
+CONF_Bool(enable_write_index_searcher_cache, "true");
+CONF_Bool(enable_index_cache_check_timestamp, "true");
+
```
Comment (xiaokang) [inline be/src/common/config.h:887]: use inverted_index_ prefix for all inverted index related config

Comment (xiaokang) [inline be/src/common/config.h:887]: use inverted_index_ prefix for all inverted index related config

File: be/src/olap/storage_engine.cpp [hunk 2]
```diff
@@ -46,6 +46,7 @@
 #include "olap/push_handler.h"
 #include "olap/reader.h"
 #include "olap/rowset/rowset_meta_manager.h"
+#include "olap/rowset/segment_v2/inverted_index_cache.h"
```
Comment (xiaokang) [inline be/src/olap/storage_engine.cpp:49]: unused include?

File: be/src/olap/lru_cache.h [hunk 3]
```diff
@@ -300,6 +302,11 @@ class HandleTable {
     void _resize();
 };
 
+typedef std::priority_queue<std::pair<int64_t, LRUHandle*>,
```
Comment (xiaokang) [inline be/src/olap/lru_cache.h:305]: add comment

File: be/src/olap/lru_cache.h [hunk 4]
```diff
@@ -353,12 +364,20 @@ class LRUCache {
 
     uint64_t _lookup_count = 0; // cache查找总次数
     uint64_t _hit_count = 0;    // 命中cache的总次数
+
+    CacheValueExtractor _cache_value_extractor;
+    bool _cache_value_check_timestamp = false;
+    LRUHandleHeap _sort_normal_entries_with_timestamp;
```
Comment (xiaokang) [inline be/src/olap/lru_cache.h:370]: sort -> sorted?

File: be/src/olap/lru_cache.h [hunk 5]
```diff
@@ -146,6 +147,7 @@ class CacheKey {
 enum class CachePriority { NORMAL = 0, DURABLE = 1 };
 
 using CacheValuePredicate = std::function<bool(const void*)>;
+using CacheValueExtractor = std::function<int64_t(const void*)>;
```
Comment (xiaokang) [inline be/src/olap/lru_cache.h:150]: add some comment

File: be/src/olap/lru_cache.cpp [hunk 6]
```diff
@@ -240,6 +240,16 @@ void LRUCache::release(Cache::Handle* handle) {
                 } else if (e->priority == CachePriority::DURABLE) {
                     _lru_append(&_lru_durable, e);
                 }
+
+                if (_cache_value_check_timestamp) {
```
Comment (xiaokang) [inline be/src/olap/lru_cache.cpp:244]: add some comment

File: be/src/olap/lru_cache.cpp [hunk 7]
```diff
@@ -250,6 +260,40 @@ void LRUCache::release(Cache::Handle* handle) {
     }
 }
 
+void LRUCache::_evict_from_lru_with_time(size_t total_size, LRUHandle** to_remove_head) {
+    // 1. evict normal cache entries
+    while (_usage + total_size > _capacity && !_sort_normal_entries_with_timestamp.empty()) {
+        auto entry_pair = _sort_normal_entries_with_timestamp.top();
+        LRUHandle* remove_handle = entry_pair.second;
+        if (_cache_value_extractor(remove_handle->value) != entry_pair.first) {
```
Comment (xiaokang) [inline be/src/olap/lru_cache.cpp:268]: add comment for this check

File: be/src/olap/lru_cache.h (hunk 5 above)
Comment (xiaokang) [reply be/src/olap/lru_cache.h:150]: Is CacheTimeExtractor more intuitive?

**Review by xiaokang — DISMISSED** [review]
LGTM, just some suggestion for minor name changes

File: be/src/common/config.h (hunk 1 above)
Comment (Tanya-W) [reply be/src/common/config.h:887]: updated

File: be/src/olap/storage_engine.cpp (hunk 2 above)
Comment (Tanya-W) [reply be/src/olap/storage_engine.cpp:49]: removed

File: be/src/olap/lru_cache.h (hunk 4 above)
Comment (Tanya-W) [reply be/src/olap/lru_cache.h:370]: updated

File: be/src/olap/lru_cache.h (hunk 3 above)
Comment (Tanya-W) [reply be/src/olap/lru_cache.h:305]: updated

File: be/src/olap/lru_cache.h (hunk 5 above)
Comment (Tanya-W) [reply be/src/olap/lru_cache.h:150]: However, it is better to name it with time first

File: be/src/olap/lru_cache.cpp (hunk 6 above)
Comment (Tanya-W) [reply be/src/olap/lru_cache.cpp:251]: updated

File: be/src/olap/lru_cache.cpp (hunk 7 above)
Comment (Tanya-W) [reply be/src/olap/lru_cache.cpp]: updated

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**